docs/proposed/hs_reassembly.md

```diff
@@ -0,0 +1,410 @@
+|TLS msg #m                            |  |TLS  #m+1                            |
+|hdr | payload                         |  |hdr | payload                        |
++----+---------------------------------+  +----+--------------------------------+
+     |  +--------+-------------------+ |       | +------------+------+--------+ |
+     |  |.. #k-1 |#k ...             | |       | |... #k      | #k+1       ...| |
+     |  +--------+-------------------+ |       | +------------+------+--------+ |
+     +---------------------------------+       +--------------------------------+
+```
+
+# Proposed solution
+
+The suggested approach is using `mbedtls_reader` to accumulate the fragments of
+incoming handshake messages. Contents of TLS records that carry handshake data
+are appended to the accumulator, and the handshake state machine consumes
+messages from the accumulator:
+
+```
+ +-------------------------+ +--------------------------+
+ |      TLS record m       | |      TLS record m+1      |
+ |+----------+ +----------+| |+------------++----------+|
+ ||   HS k   | |HS k+1 ...|| || ... HS k+1 ||HS k+2 ...||
+ |+----------+ +----------+| |+------------++----------+|
+ +-------------------------+ +--------------------------+
+                             |
+                             +------+
+                                    v
+                           +----------------+
+                           |  Accumulator   |
+                           +----------------+
+                                    |
+                                    +----------+
+                                               v
+                               +----------+ +-----------------+
+                               |   HS k   | |     HS k+1      |
+                               +----------+ +-----------------+
+
+```
+
+Conceptually, the `mbedtls_reader` is a FIFO buffer, that offers the following API:
+
+- `mbedtls_reader_feed()`    - appends data to the "tail" of the FIFO.
+- `mbedtls_reader_get()`     - provides access to the "head" of the FIFO.
+- `mbedtls_reader_commit()`  - informs the "FIFO" that the data that have been previously
+                               requested via `mbedtls_reader_get()` have been processed.
+- `mbedtls_reader_reclaim()` - cleans up the space that's occupied by the already committed
+                               messages.
+
+## Data structures
+
+Reassembly must be enabled by uncommenting `MBEDTLS_SSL_HS_REASSEMBLY`
+in the `include/mbedtls/config.h` header file, and also enabled in the runtime
+by setting a non-zero max message size.
+
+
+Suggested configuration preprocessor constructs:
+
+```
+/** Enable handshake reassembly. */
+#define MBEDTLS_SSL_HS_REASSEMBLY
+
+/** Hard limit for the maximal message size. */
+#define MBEDTLS_SSL_HS_REASSEMBLY_MAX_MESSAGE_SIZE
+
+/** The operation can not be performed because of the maximal message size limitation. */
+#define MBEDTLS_ERR_SSL_HS_REASSEMBLY_MAX_MESSAGE_SIZE_EXCEEDED
```

Comment:
Suggestion: Add a minimal example of where to call the new runtime-API, and explicitly state if anything else needs to be done by the programmer in order to make use of this feature.

Reply:
All API-related information should be contained in `ssl.h` and `config.h`, too. The implementation-details can be left to this document.